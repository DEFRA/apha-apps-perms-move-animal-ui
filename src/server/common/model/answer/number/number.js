import Joi from 'joi'
import { TextAnswer } from '../text/text.js'
import { validateAnswerAgainstSchema } from '../validation.js'

/** @import {AnswerViewModelOptions} from '../answer-model.js' */

/**
 * @param {NumberConfig} config
 * @returns {Joi.Schema}
 */
export const numberSchema = ({ payloadKey, validation }) => {
  let numberValidation = Joi.number().empty('').required()

  const messages = {
    'any.required': validation.empty?.message ?? '',
    'number.base': 'Enter a number',
    'number.integer': 'Enter a whole number',
    'number.max': validation.max?.message ?? '',
    'number.min': validation.min?.message ?? ''
  }

  if (validation.max) {
    numberValidation = numberValidation.max(validation.max.value)
  }

  if (validation.min) {
    numberValidation = numberValidation.min(validation.min.value)
  }

  return Joi.object({
    [payloadKey]: numberValidation.messages(messages).integer()
  })
}

/**
 * export @typedef {{
 *  payloadKey: string,
 *  stripWhitespace?: boolean,
 *  type?: 'email',               // the default is number, so no need to specify
 *  spellcheck?: false,
 *  characterWidth?: 2 | 10 | 20,
 *  isPageHeading? : boolean,
 *  autocomplete?: string,
 *  hint?: string,
 *  validation: {
 *    max?: { value: number, message: string },
 *    min?: { value: number, message: string },
 *    empty?: { message: string },
 *  }
 * }} NumberConfig
 */

/**
 * @typedef {string} NumberData
 */

/**
 * @template Payload
 * @augments {TextAnswer<Payload>}
 */
export class NumberAnswer extends TextAnswer {
  validate() {
    return validateAnswerAgainstSchema(
      numberSchema(this.config),
      this._data ?? {}
    )
  }
}
