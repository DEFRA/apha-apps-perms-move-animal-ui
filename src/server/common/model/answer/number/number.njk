{% from "govuk/components/input/macro.njk" import govukInput %}
{% from "govuk/components/fieldset/macro.njk" import govukFieldset %}

{% call govukFieldset() %}
  {{ govukInput(answer.viewModel(viewModelOptions)) }}
{% endcall %}
