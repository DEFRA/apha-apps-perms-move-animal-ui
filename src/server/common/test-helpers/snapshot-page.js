import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
