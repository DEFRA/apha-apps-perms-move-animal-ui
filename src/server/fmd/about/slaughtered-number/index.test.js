import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { Answer, slaughteredNumberPage } from './index.js'
import { NumberAnswer } from '~/src/server/common/model/answer/number/number.js'
import { AnimalIdsPage } from '../animal-ids/index.js'

const sectionKey = 'about'
const questionKey = 'slaughteredNumber'
const pageUrl =
  '/fmd/about-the-movement-or-activity/number-of-animals-slaughtered'
const page = slaughteredNumberPage
const question = 'How many animals will be slaughtered?'

const payload = {
  [questionKey]: 123
}

describe('Answer', () => {
  it('should be a Number input', () => {
    expect(new Answer(payload)).toBeInstanceOf(NumberAnswer)
  })

  it('should have the right payload key', () => {
    expect(Answer.config.payloadKey).toBe(questionKey)
  })

  it('should define the right empty input message', () => {
    expect(Answer.config.validation.empty?.message).toBe(
      'Enter how many animals will be slaughtered'
    )
  })

  it('should define the right minimum value and its error message', () => {
    expect(Answer.config.validation.min?.value).toBe(1)
    expect(Answer.config.validation.min?.message).toBe(
      'Enter a number 1 or above'
    )
  })

  it('should define the right maximum value and its error message', () => {
    expect(Answer.config.validation.max?.value).toBe(5000)
    expect(Answer.config.validation.max?.message).toBe(
      'Enter a number 5000 or below'
    )
  })
})

describe('SlaughteredNumberPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  describe('nextPage', () => {
    it('should return animalIdsPage for any value', () => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(AnimalIdsPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
