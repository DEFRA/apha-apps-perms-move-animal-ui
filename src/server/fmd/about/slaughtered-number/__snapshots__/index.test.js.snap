// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlaughteredNumberPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="slaughteredNumber">
      How many animals will be slaughtered?
    </label>
  </h1>
  <input class="govuk-input govuk-input--width-10" id="slaughteredNumber" name="slaughteredNumber" type="text">
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
